### PR TITLE
Add testing for regex matching and other custom kwargs

### PR DESCRIPTION
**Pull request by SumanthRH**
# What does this PR do?

Currently, testing wrt different types of config kwargs is a bit disorganized. This is my attempt to have a file where custom kwargs for different adapters can be added. For example, I have added a test against replacing all T5 linear layers with LoRA layers. There was an issue before with the regex matching #778 that made it work for value but not key layers, so I'm testing if all the layers are replaced. Also with IA3, there are currently no tests that check whether the right layers have been marked as feedforward (and if other layers are untouched). I think one can add other tests here for AdaLoRA, etc. as well. 

**Does this improve test coverage?** Not in terms of lines of code. There are some string inputs for `target_modules` in `test_custom_models.py` that end up testing the regex matching logic. I think having a separate file like this is still useful for other behaviour (like the corner case with T5 key layers, or feedforward with IA3).

### CODE REVIEW

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for extending the test coverage when it comes to matching layer names.

I haven't done a deep dive into the code, but wanted to bring up an idea and hear your opinion. In an older PR, I already tried to work on this type of tests: #690. It was never merged, but the idea was to factor out the function responsible for matching the layer name and testing only that function specifically.

From my point of view, the advantage is that the test is much more specific, we don't have to load an actual model and initialize it, we just test the matching. Now I wonder if we can still do this type of refactor to create a standalone matching function that takes the config arguments and a layer name, then just returns True or False, like `_is_matching_module_name` in that PR.

We could still have tests similar to what is proposed in this PR, which are more end-to-end, but they could be much simpler, since we would already have tests for the layer name matching and could skip that part.

**Comment by SumanthRH** [comment]
@BenjaminBossan I went over #690 and have done similar refactoring: added a `inspect_matched_modules()` to the base tuner class, and then adde your list of tests - which test the `check_target_module_exists` function. Currently, there's one corner case that I'm not sure of - when `layers_pattern` is an empty list,  `check_target_module_exists` will ignore layers_to_transform and thus all layer indexes get matched. I feel that `layers_pattern=[]` and `layers_pattern=None` should give the same result (the other option is to not match any layer at all, which, doesn't make any sense :p). I can add the changes to `check_target_module_exists` if you think this is the right behaviour. 

Also, for LoRAConfig,  if `target_modules` is a regex string, then `layers_to_transform` and `layers_pattern` will get ignored in `check_target_module_exists`. I think it's better to enforce that both arguments are None, and raise a valueError if otherwise. I've made that change and added a test in `test_config.py`

File: src/peft/tuners/lora/config.py [hunk 1]
```diff
@@ -88,10 +88,11 @@ class LoraConfig(PeftConfig):
     layers_to_transform: Optional[Union[List[int], int]] = field(
         default=None,
         metadata={
-            "help": "The layer indexes to transform, is this argument is specified, PEFT will transform only the layers indexes that are specified inside this list. If a single integer is passed, PEFT will transform only the layer at this index."
+            "help": "The layer indexes to transform, is this argument is specified, PEFT will transform only the layers indexes that are specified inside this list. If a single integer is passed, PEFT will transform only the layer at this index. "
+            "This cannot be used when `target_modules` is a regex expression"
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/config.py:92]: How about, instead of wording this in the negative ("cannot be used"), we word it positively: "This only works when `target_modules` is a list of str." or something like that. Should make it easier for users to know what they need to do.

File: src/peft/tuners/lora/config.py [hunk 2]
```diff
@@ -121,3 +122,10 @@ def __post_init__(self):
         self.target_modules = (
             set(self.target_modules) if isinstance(self.target_modules, list) else self.target_modules
         )
+        # if target_modules is a regex expression, then layers_to_transform should be None
+        if isinstance(self.target_modules, str) and self.layers_to_transform is not None:
+            raise ValueError("`layers_to_transform` cannot be used when `target_modules` is a regex expression.")
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/config.py:127]:
```suggestion
            raise ValueError("`layers_to_transform` cannot be used when `target_modules` is str.")
```

I think this is better, because the user might not be aware that the str they are passing is used as a regex.

File: src/peft/tuners/lora/config.py [hunk 3]
```diff
@@ -121,3 +122,10 @@ def __post_init__(self):
...
+
+        # if target_modules is a regex expression, then layers_pattern should be None
+        if isinstance(self.target_modules, str) and self.layers_pattern is not None:
+            raise ValueError("`layers_pattern` cannot be used when `target_modules` is a regex expression.")
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/config.py:131]:
```suggestion
            raise ValueError("`layers_pattern` cannot be used when `target_modules` is a str.")
```

Same argument.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for working on this. I have some smaller comments about wording, please take a look.

Then I have a question regarding `inspect_matched_modules`: AFAICT, it is not actually being used in the code, only for the tests. Therefore, I would not add it to the base tuner class. It could instead be a helper function that is defined in the test file.

The tests could also be extended to cover other adapter types than LoRA, but it would be fine to add that later.

> Currently, there's one corner case that I'm not sure of - when layers_pattern is an empty list, check_target_module_exists will ignore layers_to_transform and thus all layer indexes get matched. I feel that layers_pattern=[] and layers_pattern=None should give the same result (the other option is to not match any layer at all, which, doesn't make any sense :p). I can add the changes to check_target_module_exists if you think this is the right behaviour.

Fair point. For this PR, I would not change the logic, it's easier if we deal with it in a separate PR if we think it's a good change.

> Also, for LoRAConfig, if target_modules is a regex string, then layers_to_transform and layers_pattern will get ignored in check_target_module_exists. I think it's better to enforce that both arguments are None, and raise a valueError if otherwise. I've made that change and added a test in test_config.py

Also good point. I wonder if we should issue a warning instead, as it could theoretically break existing code, although arguably, that code would already not be doing what it's supposed to do. I can live with either solution, just wanted highlight this option.

**Comment by SumanthRH** [comment]
> AFAICT, it is not actually being used in the code, only for the tests. Therefore, I would not add it to the base tuner class. It could instead be a helper function that is defined in the test file

Good idea! I've kept it to be a helper function in `tuners_utils.py` itself. Perhaps it can be used in case a user wants to do this as well. 

>  I wonder if we should issue a warning instead, as it could theoretically break existing code, although arguably, that code would already not be doing what it's supposed to do. I can live with either solution, just wanted highlight this option.

Noted. I think here it is better left as an explicit error, because it might be easy to make a mistake and use `target_modules="baz"` instead of `target_modules=["baz"]`. In the first case, `layers_to_transform` and `layers_pattern` get ignored and warnings usually get lost. Usually, the typical stack with PEFT training is NCCL + Pytorch + Transformers + Accelerate + PEFT . If we're doing multi-gpu training, we have debug messages and warnings from all of these libraries, logged per gpu, and my experience has been that it's too painful to keep track of warnings!

**Review by BenjaminBossan — COMMENTED** [review]
This looks pretty good, thanks for your work on improving the test coverage when it comes to matching module names. Probably there is some work we can do in the future to streamline the logic of module matching, and this PR gives us confidence that we won't break anything important.

One last tiny request I have is to change the name of `test_custom_kwargs.py`. This name isn't very descriptive. How about `test_tuners_utils.py` instead?

**Comment by SumanthRH** [comment]
Good idea! I've renamed the file @BenjaminBossan !

**Review by BenjaminBossan — APPROVED** [review]
Thanks for this PR, it's always good to improve test coverage.
